x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -1053,63 +1058,84 @@ public void expandIds(
     }
 
     public void getInferenceStats(String[] modelIds, @Nullable TaskId parentTaskId, ActionListener<List<InferenceStats>> listener) {
-        MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
-        Arrays.stream(modelIds).map(TrainedModelProvider::buildStatsSearchRequest).forEach(multiSearchRequest::add);
-        if (multiSearchRequest.requests().isEmpty()) {
-            listener.onResponse(Collections.emptyList());
-            return;
-        }
-        if (parentTaskId != null) {
-            multiSearchRequest.setParentTask(parentTaskId);
-        }
-        executeAsyncWithOrigin(
-            client.threadPool().getThreadContext(),
-            ML_ORIGIN,
-            multiSearchRequest,
-            ActionListener.<MultiSearchResponse>wrap(responses -> {
-                List<InferenceStats> allStats = new ArrayList<>(modelIds.length);
-                int modelIndex = 0;
-                assert responses.getResponses().length == modelIds.length : "mismatch between search response size and models requested";
-                for (MultiSearchResponse.Item response : responses.getResponses()) {
-                    if (response.isFailure()) {
-                        if (ExceptionsHelper.unwrapCause(response.getFailure()) instanceof ResourceNotFoundException) {
-                            modelIndex++;
-                            continue;
+
+        SubscribableListener.<ClusterHealthResponse>newForked((delegate) -> {
+            // first wait for the index to be available
+            executeAsyncWithOrigin(
+                client.threadPool().getThreadContext(),
+                ML_ORIGIN,
+                new ClusterHealthRequest(new TimeValue(2, TimeUnit.SECONDS), MlStatsIndex.indexPattern()).waitForYellowStatus(),
+                delegate,
+                client.admin().cluster()::health
+            );
+        }).<List<InferenceStats>>andThen((delegate, clusterHealthResponse) -> {
```

Comment:
The provided function executes on the thread handling the ClusterHealth response, `management` or `transport`.  We should probably move it back to ML?

```suggestion
        })
            .<List<InferenceStats>>andThen(
                client.threadPool().executor(MachineLearning.UTILITY_THREAD_POOL_NAME),
                client.threadPool().getThreadContext(),
                (delegate, clusterHealthResponse) -> {
```

If we don't want that ridiculous indenting, we can maybe create methods to shorten it?

```java
    private Executor mlExecutor() {
        return client.threadPool().executor(MachineLearning.UTILITY_THREAD_POOL_NAME);
    }

    private ThreadContext threadContext() {
        return client.threadPool().getThreadContext();
    }
```

Reply:
good catch!

Reply:
thinking about this more, I don't think that executing this code in a threadpool makes much of a difference because the code being executed here is pretty trivial and the request executions are async. regardless, I'll put it on a threadpool because I don't think theres any harm in that.